modules/metastorage/src/integrationTest/java/org/apache/ignite/internal/metastorage/impl/ItIdempotentCommandCacheTest.java

```diff
@@ -0,0 +1,224 @@
+ */
+@ExtendWith(ConfigurationExtension.class)
+public class ItIdempotentCommandCacheTest extends IgniteAbstractTest {
+    private ClusterService clusterService;
+
+    private Loza raftManager;
+
+    private KeyValueStorage storage;
+
+    private MetaStorageManagerImpl metaStorageManager;
+
+    @BeforeEach
+    void setUp(
+            TestInfo testInfo,
+            @InjectConfiguration("mock.responseTimeout = 100") RaftConfiguration raftConfiguration,
+            @InjectConfiguration("mock.idleSyncTimeInterval = 100") MetaStorageConfiguration metaStorageConfiguration
+    ) {
+        var addr = new NetworkAddress("localhost", 10_000);
+
+        clusterService = clusterService(testInfo, addr.port(), new StaticNodeFinder(List.of(addr)));
+
+        HybridClock clock = new HybridClockImpl();
+
+        var raftGroupEventsClientListener = new RaftGroupEventsClientListener();
+
+        raftManager = new Loza(clusterService, new NoOpMetricManager(), raftConfiguration, workDir.resolve("loza"), clock,
+                raftGroupEventsClientListener);
+
+        var logicalTopologyService = mock(LogicalTopologyService.class);
+
+        var topologyAwareRaftGroupServiceFactory = new TopologyAwareRaftGroupServiceFactory(
+                clusterService,
+                logicalTopologyService,
+                Loza.FACTORY,
+                raftGroupEventsClientListener
+        );
+
+        ClusterManagementGroupManager cmgManager = mock(ClusterManagementGroupManager.class);
+
+        when(cmgManager.metaStorageNodes()).thenReturn(completedFuture(Set.of(clusterService.nodeName())));
+
+        storage = new RocksDbKeyValueStorage(
+                clusterService.nodeName(),
+                workDir.resolve("metastorage"),
+                new NoOpFailureProcessor(clusterService.nodeName()));
+
+        metaStorageManager = new MetaStorageManagerImpl(
+                clusterService,
+                cmgManager,
+                logicalTopologyService,
+                raftManager,
+                storage,
+                clock,
+                topologyAwareRaftGroupServiceFactory,
+                metaStorageConfiguration
+        );
+
+        assertThat(
+                startAsync(clusterService, raftManager, metaStorageManager)
+                        .thenCompose(unused -> metaStorageManager.deployWatches()),
+                willCompleteSuccessfully()
+        );
+    }
+
+    @AfterEach
+    void tearDown() throws Exception {
+        List<IgniteComponent> components = List.of(metaStorageManager, raftManager, clusterService);
+
+        closeAll(Stream.concat(
+                components.stream().map(c -> c::beforeNodeStop),
+                Stream.of(() -> assertThat(stopAsync(components), willCompleteSuccessfully()))
+        ));
+    }
+
+    @Test
+    public void testIdempotentInvoke() throws InterruptedException {
+        ByteArray testKey = new ByteArray("key".getBytes(StandardCharsets.UTF_8));
```

Comment:
Could you please also add a test that will verify that idempotent command cache is available (and thus proper command result is returned) after leader change?

Reply:
added ItIdempotentCommandCacheTest#testIdempotentInvokeAfterLeaderChange